
<launch>

      <!-- global path planner name -->
      <arg name="global_planner" default="astar" />
      <!-- local planner name -->
      <arg name="local_planner" default="dwa" />

      <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" >
        <!-- <param name="base_global_planner" value="path_planner/PathPlanner" /> -->
        <param name="base_global_planner" value="navfn/NavfnROS" />
        <param name="PathPlanner/planner_name" value="$(arg global_planner)" />

                <!-- graph search -->
        <rosparam file="$(find me5413_navigation)/config/planner/common_planner_params.yaml" command="load" />
        <rosparam file="$(find me5413_navigation)/config/planner/graph_planner_params.yaml" command="load"
            if="$(eval arg('global_planner')=='astar'
                    or arg('global_planner')=='jps' 
                    or arg('global_planner')=='bi_jps' 
                    or arg('global_planner')=='gbfs'
                    or arg('global_planner')=='dijkstra'
                    or arg('global_planner')=='dstar'
                    or arg('global_planner')=='lpa_star'
                    or arg('global_planner')=='voronoi'
                    or arg('global_planner')=='dstar_lite'
                    or arg('global_planner')=='theta_star'
                    or arg('global_planner')=='lazy_theta_star'
                    or arg('global_planner')=='s_theta_star'
                    or arg('global_planner')=='hybrid_astar'
                    or arg('global_planner')=='lazy'
                )" />

        <!-- sample search -->
        <rosparam file="$(find me5413_navigation)/config/planner/sample_planner_params.yaml" command="load"
            if="$(eval arg('global_planner')=='rrt'
                    or arg('global_planner')=='rrt_star'
                    or arg('global_planner')=='informed_rrt'
                    or arg('global_planner')=='quick_informed_rrt'
                    or arg('global_planner')=='rrt_connect'
                )" /> 

        <!-- evolutionary search -->
        <rosparam file="$(find me5413_navigation)/config/planner/evolutionary_planner_params.yaml" command="load"
            if="$(eval arg('global_planner')=='aco'
                    or arg('global_planner')=='pso'
                    or arg('global_planner')=='ga'
                )" />

        <!-- local planner plugin -->
        <param name="base_local_planner" value="dwa_controller/DWAController" if="$(eval arg('local_planner')=='dwa')" />
        <rosparam file="$(find me5413_navigation)/config/controller/dwa_controller_params.yaml" command="load" if="$(eval arg('local_planner')=='dwa')" />

        <param name="base_local_planner" value="pid_controller/PIDController" if="$(eval arg('local_planner')=='pid')" />
        <rosparam file="$(find me5413_navigation)/config/controller/pid_controller_params.yaml" command="load" if="$(eval arg('local_planner')=='pid')" />

        <param name="base_local_planner" value="apf_controller/APFController" if="$(eval arg('local_planner')=='apf')" />
        <rosparam file="$(find me5413_navigation)/config/controller/apf_controller_params.yaml" command="load" if="$(eval arg('local_planner')=='apf')" />

        <param name="base_local_planner" value="rpp_controller/RPPController" if="$(eval arg('local_planner')=='rpp')" />
        <rosparam file="$(find me5413_navigation)/config/controller/rpp_controller_params.yaml" command="load" if="$(eval arg('local_planner')=='rpp')" />

        <param name="base_local_planner" value="lqr_controller/LQRController" if="$(eval arg('local_planner')=='lqr')" />
        <rosparam file="$(find me5413_navigation)/config/controller/lqr_controller_params.yaml" command="load" if="$(eval arg('local_planner')=='lqr')" />

        <param name="base_local_planner" value="mpc_controller/MPCController" if="$(eval arg('local_planner')=='mpc')" />
        <rosparam file="$(find me5413_navigation)/config/controller/mpc_controller_params.yaml" command="load" if="$(eval arg('local_planner')=='mpc')" />

        <param name="base_local_planner" value="static_controller/StaticController" if="$(eval arg('local_planner')=='static')" />

        <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" if="$(eval arg('local_planner')=='teb')" />
        <rosparam file="$(find me5413_navigation)/config/controller/teb_local_planner_params.yaml" command="load" if="$(eval arg('local_planner')=='teb')"/>
        <!-- ORCA -->
        <!-- <param name="base_local_planner" value="orca_controller/ORCAController" if="$(eval arg('local_planner')=='orca')" />
        <param name="ORCAController/agent_number" value="$(arg agent_number)" if="$(eval arg('local_planner')=='orca')" />
        <param name="ORCAController/agent_id" value="$(arg agent_id)" if="$(eval arg('local_planner')=='orca')" />
        <rosparam file="$(find me5413_navigation)/config/controller/orca_controller_params.yaml" command="load" if="$(eval arg('local_planner')=='orca')" /> -->


      <rosparam file="$(find me5413_navigation)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find me5413_navigation)/config/costmap_common_params.yaml" command="load" ns="local_costmap" />
      
      <rosparam file="$(find me5413_navigation)/config/map_nav_params/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find me5413_navigation)/config/map_nav_params/global_costmap_param_teb.yaml" command="load" />
      
      
      <rosparam file="$(find me5413_navigation)/config/base_local_planner_params.yaml" command="load" /> 
      <rosparam file="$(find me5413_navigation)/config/move_base_params.yaml" command="load" />
      
      <!-- Two additional yml modification -->
      <rosparam file="$(find me5413_navigation)/config/addition_params/move_base.yaml" command="load" />
      <rosparam file="$(find me5413_navigation)/config/addition_params/global_planner.yaml" command="load" />

      <!-- <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS"/> -->
      
      <remap from="odom" to="odometry/filtered" />

    </node>

        <!-- <node pkg="topic_tools" type="relay" name="relay_ndt_to_amcl" args="/ndt/pose /amcl_pose" /> -->

  </launch>